Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrBuilder.kt

```diff
@@ -0,0 +1,352 @@
+
+        override fun getKotlinNameForValue(enumDef: EnumDef): String = enumDef.kotlinName
+
+        override fun getPackageFor(declaration: TypeDeclaration): String {
+            return imports.getPackage(declaration.location) ?: pkgName
+        }
+
+        override val useUnsignedTypes: Boolean
+            get() = when (platform) {
+                KotlinPlatform.JVM -> false
+                KotlinPlatform.NATIVE -> true
+            }
+    }
+
+    override val macroConstantsByName: Map<String, MacroDef> =
+            (nativeIndex.macroConstants + nativeIndex.wrappedMacros).associateBy { it.name }
+
+    /**
+     * The name to be used for this enum in Kotlin
+     */
+    val EnumDef.kotlinName: String
+        get() = if (spelling.startsWith("enum ")) {
+            spelling.substringAfter(' ')
+        } else {
+            assert (!isAnonymous)
+            spelling
+        }
+
+
+    private val pkgName: String
+        get() = configuration.pkgName
+
+    /**
+     * The name to be used for this struct in Kotlin
+     */
+    val StructDecl.kotlinName: String
+        get() = stubIrContext.getKotlinName(this)
+
+    override fun tryCreateIntegralStub(type: Type, value: Long): IntegralConstantStub? {
+        val integerType = type.unwrapTypedefs() as? IntegerType ?: return null
+        val size = integerType.size
+        if (size != 1 && size != 2 && size != 4 && size != 8) return null
+        return IntegralConstantStub(value, size, declarationMapper.isMappedToSigned(integerType))
+    }
+
+    override fun tryCreateDoubleStub(type: Type, value: Double): DoubleConstantStub? {
+        val unwrappedType = type.unwrapTypedefs() as? FloatingType ?: return null
+        val size = unwrappedType.size
+        if (size != 4 && size != 8) return null
+        return DoubleConstantStub(value, size)
+    }
+
+    override val bridgeComponentsBuilder = BridgeGenerationComponentsBuilder()
+
+    override fun getKotlinClassFor(objCClassOrProtocol: ObjCClassOrProtocol, isMeta: Boolean): Classifier {
+        return declarationMapper.getKotlinClassFor(objCClassOrProtocol, isMeta)
+    }
+
+    override fun getKotlinClassForPointed(structDecl: StructDecl): Classifier {
+        val classifier = declarationMapper.getKotlinClassForPointed(structDecl)
+        return classifier
+    }
+}
+
+data class StubIrBuilderResult(
+        val stubs: SimpleStubContainer,
+        val declarationMapper: DeclarationMapper,
+        val bridgeGenerationComponents: BridgeGenerationComponents
+)
+
+/**
+ * Produces [StubIrBuilderResult] for given [KotlinPlatform] using [InteropConfiguration].
+ */
+class StubIrBuilder(
+        private val context: StubIrContext,
+        private val verbose: Boolean = false
+) {
+
+    private val configuration = context.configuration
+    private val nativeIndex: NativeIndex = context.nativeIndex
+
+    private val classes = mutableListOf<ClassStub>()
+    private val functions = mutableListOf<FunctionalStub>()
```

Comment:
Is `FunctionalStub` really required here instead of `FunctionStub`?

Reply:
Nope, it's a leftover from a refactoring.